File: backend/app/schemas/metajungle.py
"""Meta-Jungle API schemas."""
import uuid
from datetime import datetime
from enum import Enum
from typing import Optional, List, Any
from pydantic import BaseModel, Field, ConfigDict


# ── Reputation (Chapter 6) ──────────────────────────────────────────────────
class ReputationResponse(BaseModel):
    user_id: uuid.UUID
    activity_score: int = Field(..., ge=0, le=1000)
    reputation_score: int = Field(..., ge=0, le=1000)
    influence_score: int = Field(..., ge=0, le=1000)
    role: str
    earn_multiplier: float
    next_role: Optional[str] = None


# ── Quests (Chapter 5) ──────────────────────────────────────────────────────
class QuestResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    description: str
    pp_reward: int
    category: str
    verification_type: str
    min_role: str
    steps: Optional[Any] = None
    daily_limit: int
    action_url: Optional[str] = None
    is_active: bool
    starts_at: Optional[datetime] = None
    ends_at: Optional[datetime] = None


class QuestListResponse(BaseModel):
    quests: List[QuestResponse]
    total: int


class QuestCompleteRequest(BaseModel):
    proof: Optional[dict] = Field(default=None, description="Verification proof payload")


class QuestCompletionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    quest_id: uuid.UUID
    status: str
    pp_awarded: float
    proof: Optional[Any] = None
    created_at: datetime


# ── NFT Vault (Chapter 4.7) ─────────────────────────────────────────────────
class NFTResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    name: str
    tier: str
    daily_pp_rate: int
    contract_address: str
    token_id: str
    traits: Optional[Any] = None
    utilities: Optional[Any] = None
    is_staked: bool


class NFTListResponse(BaseModel):
    nfts: List[NFTResponse]
    total: int
    total_daily_yield: int


# ── P2P (Chapter 4.5) ───────────────────────────────────────────────────────
class P2POrderResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    side: str
    pp_amount: int
    price: float
    currency: str
    payment_method: str
    status: str
    seller_id: Optional[uuid.UUID] = None
    buyer_id: Optional[uuid.UUID] = None
    created_at: datetime


class P2POrderListResponse(BaseModel):
    orders: List[P2POrderResponse]
    total: int


class P2POrderCreate(BaseModel):
    side: str = Field(..., pattern="^(buy|sell)$")
    pp_amount: int = Field(..., gt=0)
    price: float = Field(..., gt=0)
    currency: str = Field(default="NGN", max_length=8)
    payment_method: str = Field(..., min_length=1, max_length=64)


# ── Staking (Chapter 7) ─────────────────────────────────────────────────────
class StakeResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    asset: str
    pp_amount: int
    multiplier: float
    lock_days: int
    accrued: float
    status: str
    started_at: datetime
    unlocked_at: Optional[datetime] = None


class StakeListResponse(BaseModel):
    stakes: List[StakeResponse]
    total_staked: int
    total_accrued: float


class StakeCreate(BaseModel):
    pp_amount: int = Field(..., gt=0)
    lock_days: int = Field(..., description="One of 30, 90, 180")


# ── Campaigns (Chapter 11) ──────────────────────────────────────────────────
class CampaignResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    brand: Optional[str] = None
    title: str
    blurb: str
    pp_budget: int
    pp_per_task: int
    pp_claimed: int
    status: str
    featured: bool
    total_participants: int
    ends_at: Optional[datetime] = None


class CampaignListResponse(BaseModel):
    campaigns: List[CampaignResponse]
    total: int


# ── Learn (Chapter 13) ──────────────────────────────────────────────────────
class CourseResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    title: str
    blurb: str
    level: str
    lessons: int
    pp_reward: int


class CourseListResponse(BaseModel):
    courses: List[CourseResponse]
    total: int


class QuizSubmitRequest(BaseModel):
    answers: List[int] = Field(..., description="Selected option index per question")


class QuizResultResponse(BaseModel):
    passed: bool
    score: int
    pp_awarded: float


# ── Marketplace (Chapter 12) ────────────────────────────────────────────────
class RedeemRequest(BaseModel):
    product_id: str = Field(..., min_length=1)
    destination: Optional[str] = Field(default=None, description="Phone, meter, email, etc.")


class RedemptionResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    product_id: str
    product_name: str
    category: str
    pp_cost: int
    destination: Optional[str] = None
    voucher_code: str
    status: str
    created_at: datetime


class MarketProduct(BaseModel):
    id: str
    category: str
    name: str
    pp: int
    fiat: str
    provider: str
    regions: List[str]
    input: str


class MarketCatalogResponse(BaseModel):
    products: List[MarketProduct]

File: apps/web/src/api/metajungle.ts
/**
 * Meta-Jungle ecosystem API client (reputation, quests, NFT, P2P, staking,
 * campaigns, learn, marketplace). Backed by the FastAPI endpoints documented
 * in backend/METAJUNGLE_API.md.
 */
import apiClient from '@/lib/api';

// ── Reputation ──────────────────────────────────────────────────────────────
export interface Reputation {
  user_id: string;
  activity_score: number;
  reputation_score: number;
  influence_score: number;
  role: string;
  earn_multiplier: number;
  next_role: string | null;
}

// ── Quests ──────────────────────────────────────────────────────────────────
export interface ApiQuest {
  id: string;
  title: string;
  description: string;
  pp_reward: number;
  category: string;
  verification_type: string;
  min_role: string;
  steps?: { label: string; verification: string }[] | null;
  daily_limit: number;
  action_url?: string | null;
  screenshot_required: boolean;
  link_required: boolean;
  is_active: boolean;
  starts_at?: string | null;
  ends_at?: string | null;
}

// ── NFT ─────────────────────────────────────────────────────────────────────
export interface ApiNFT {
  id: string;
  name: string;
  tier: 'common' | 'rare' | 'epic' | 'legendary';
  daily_pp_rate: number;
  contract_address: string;
  token_id: string;
  traits?: { label: string; value: string }[] | null;
  utilities?: string[] | null;
  is_staked: boolean;
}

// ── P2P ─────────────────────────────────────────────────────────────────────
export interface ApiP2POrder {
  id: string;
  side: 'buy' | 'sell';
  pp_amount: number;
  price: number;
  currency: string;
  payment_method: string;
  status: string;
  seller_id?: string | null;
  buyer_id?: string | null;
  created_at: string;
}

// ── Staking ─────────────────────────────────────────────────────────────────
export interface ApiStake {
  id: string;
  asset: string;
  pp_amount: number;
  multiplier: number;
  lock_days: number;
  accrued: number;
  status: string;
  started_at: string;
  unlocked_at?: string | null;
}

// ── Campaigns ───────────────────────────────────────────────────────────────
export interface ApiCampaign {
  id: string;
  slug: string;
  brand?: string | null;
  partner_tier?: string | null;
  title: string;
  blurb: string;
  pp_budget: number;
  pp_per_task: number;
  pp_claimed: number;
  pp_reserved: number;
  pp_available: number;
  status: 'draft' | 'active' | 'paused' | 'ended';
  featured: boolean;
  total_participants: number;
  max_participants?: number | null;
  target_regions: string[];
  target_roles: string[];
  min_role: string;
  joined: boolean;
  my_pp_earned: number;
  my_pp_withdrawn: number;
  my_pp_available: number;
  starts_at?: string | null;
  ends_at?: string | null;
}

export type ApiVerificationType =
  | 'oauth'
  | 'webhook'
  | 'manual'
  | 'screenshot'
  | 'on_chain';

export interface ApiCampaignTask {
  id: string;
  campaign_id: string;
  title: string;
  description: string;
  pp_reward: number;
  verification_type: ApiVerificationType;
  daily_limit: number;
  action_url?: string | null;
  screenshot_required: boolean;
  link_required: boolean;
  order_index: number;
  is_active: boolean;
  completed_today: number;
  can_complete: boolean;
}

export interface ApiCampaignEligibility {
  eligible: boolean;
  reason?: string | null;
  role: string;
  earn_multiplier: number;
}

export interface ApiCampaignCompletion {
  id: string;
  campaign_id: string;
  task_id: string;
  status: 'pending' | 'approved' | 'rejected';
  pp_awarded: number;
  created_at: string;
}

// ── Learn ───────────────────────────────────────────────────────────────────
export interface ApiCourse {
  id: string;
  title: string;
  blurb: string;
  level: 'Beginner' | 'Intermediate' | 'Advanced';
  lessons: number;
  pp_reward: number;
  quiz?: { q: string; options: string[] } | null;
}

// ── Marketplace ─────────────────────────────────────────────────────────────
export interface ApiProduct {
  id: string;
  category: string;
  name: string;
  pp: number;
  fiat: string;
  provider: string;
  regions: string[];
  input: string;
}

export interface ApiRedemption {
  id: string;
  product_id: string;
  product_name: string;
  category: string;
  pp_cost: number;
  destination?: string | null;
  voucher_code: string;
  status: string;
  created_at: string;
}

export const metajungleAPI = {
  getReputation: async (): Promise<Reputation> =>
    (await apiClient.get('/reputation/me')).data,

  listQuests: async (): Promise<ApiQuest[]> =>
    (await apiClient.get('/quests')).data.quests,
  completeQuest: async (id: string, proof?: Record<string, unknown>) =>
    (await apiClient.post(`/quests/${id}/complete`, { proof: proof ?? null })).data,
  getMyCompletions: async (): Promise<Record<string, string>> =>
    (await apiClient.get('/quests/my-completions')).data,

  listNFTs: async (): Promise<{ nfts: ApiNFT[]; total_daily_yield: number }> =>
    (await apiClient.get('/nft')).data,

  listOrders: async (side?: 'buy' | 'sell'): Promise<ApiP2POrder[]> =>
    (await apiClient.get('/p2p/orders', { params: side ? { side } : {} })).data.orders,
  createOrder: async (body: {
    side: 'buy' | 'sell';
    pp_amount: number;
    price: number;
    currency?: string;
    payment_method: string;
  }) => (await apiClient.post('/p2p/orders', body)).data,

  listStakes: async (): Promise<{ stakes: ApiStake[]; total_staked: number; total_accrued: number }> =>
    (await apiClient.get('/staking')).data,
  createStake: async (pp_amount: number, lock_days: number) =>
    (await apiClient.post('/staking', { pp_amount, lock_days })).data,
  claimStake: async (id: string) =>
    (await apiClient.post(`/staking/${id}/claim`)).data,
  unstake: async (id: string) =>
    (await apiClient.post(`/staking/${id}/unstake`)).data,

  listCampaigns: async (): Promise<ApiCampaign[]> =>
    (await apiClient.get('/campaigns')).data.campaigns,
  joinCampaign: async (id: string) =>
    (await apiClient.post(`/campaigns/${id}/join`)).data,
  listCampaignTasks: async (id: string): Promise<ApiCampaignTask[]> =>
    (await apiClient.get(`/campaigns/${id}/tasks`)).data.tasks,
  campaignEligibility: async (id: string): Promise<ApiCampaignEligibility> =>
    (await apiClient.get(`/campaigns/${id}/eligibility`)).data,
  completeCampaignTask: async (
    campaignId: string,
    taskId: string,
    proof?: Record<string, unknown>,
  ): Promise<ApiCampaignCompletion> =>
    (await apiClient.post(`/campaigns/${campaignId}/tasks/${taskId}/complete`, { proof }))
      .data,
  withdrawCampaignPoints: async (campaignId: string) =>
    (await apiClient.post(`/campaigns/${campaignId}/withdraw`)).data,

  listCourses: async (): Promise<ApiCourse[]> =>
    (await apiClient.get('/learn/courses')).data.courses,
  submitQuiz: async (id: string, answers: number[]): Promise<{ passed: boolean; score: number; pp_awarded: number }> =>
    (await apiClient.post(`/learn/courses/${id}/quiz`, { answers })).data,

  getCatalog: async (): Promise<ApiProduct[]> =>
    (await apiClient.get('/marketplace/catalog')).data.products,
  redeem: async (product_id: string, destination?: string): Promise<ApiRedemption> =>
    (await apiClient.post('/marketplace/redeem', { product_id, destination: destination ?? null })).data,
};
